docs/book/v3/reference/migration.md

```diff
@@ -0,0 +1,135 @@
+- [laminas/laminas-cache-storage-adapter-filesystem](https://github.com/laminas/laminas-cache-storage-adapter-filesystem) 
+- [laminas/laminas-cache-storage-adapter-memcached](https://github.com/laminas/laminas-cache-storage-adapter-memcached)
+- [laminas/laminas-cache-storage-adapter-memory](https://github.com/laminas/laminas-cache-storage-adapter-memory)
+- [laminas/laminas-cache-storage-adapter-redis](https://github.com/laminas/laminas-cache-storage-adapter-redis) 
+- [laminas/laminas-cache-storage-adapter-session](https://github.com/laminas/laminas-cache-storage-adapter-session)
+
+# Composition Over Inheritance
+
+In case you are extending one of the cache implementations, your code might look as follows:
+
+
+```php
+use Laminas\Cache\Storage\Adapter\Filesystem;
+
+class MyFileystemStorage extends Filesystem
+{
+    protected function internalSetItem(&$normalizedKey,&$value)
+    {
+        $value = doSomethingWithValue($value);
+        $normalizedKey = doSOmethingWithKey($normalizedKey);
+        return parent::internalSetItem($normalizedKey,$value);
+    }
+} 
+```
+
+If this looks familiar, using composition would look like this:
+
+```php
+use Laminas\Cache\Exception;
+use Laminas\Cache\Storage\Adapter\AbstractAdapter;
+use Laminas\Cache\Storage\Adapter\Filesystem;
+use Laminas\Cache\Storage\Adapter\FilesystemOptions;
+
+final class MyFilesytemStorage extends AbstractAdapter
+{
+    /**
+    * @var \Laminas\Cache\Storage\Adapter\Filesystem
+    */
+    private $filesystem;
+    
+    public function __construct(Filesystem $filesystem) 
+    {
+        $this->filesystem = $filesystem;
+        parent::__construct();
+    }
+    
+    protected function internalGetItem(&$normalizedKey,&$success = null,&$casToken = null)
+    {
+        return $this->filesystem->getItem(
+            $normalizedKey, 
+            $success,
+            $casToken
+        );
+    }
+    
+    protected function internalSetItem(&$normalizedKey,&$value)
+    {
+        $value = doSomethingWithValue($value);
+        $normalizedKey = doSomethingWithValue($normalizedKey);
+        return $this->filesystem->setItem(
+            $normalizedKey, 
+            $value
+        );
+    }
+    
+    protected function internalRemoveItem(&$normalizedKey)
+    {
+        return $this->filesystem->removeItem(
+            $normalizedKey
+        );
+    }   
+}
+```
+
+Even tho, that this is more code to add/change than the previous solution, this gives the maintainers of the cache adapters more freedom to provide more stable adapter implementations.
+
+If this does not fit your requirements, please let us know via the `laminas-cache` [discussions on github](https://github.com/laminas/laminas-cache/discussions) and tell us more about your implementations. Maybe your addition should be part of our official adapter or could be provided as a dedicated Plugin instead.
```

Comment:
So far, we do not officially use the discussions. I think this should be discussed in a TSC meeting.
(The biggest problem for me is that the tab/link for discussions is [missing on the organisation page](https://github.com/github/feedback/discussions/4031).)

Reply:
Sure, but I'd prefer having a discussion for migration issues rather than issues which are used to discuss solutions.
But I'd be fine with removing this and disable the discussion section for this component if this is a showstopper.

Reply:
Don't get me wrong, I hope that in the future we can have all discussions here on GitHub.

Therefore, I suggest the following:

```suggestion
If this does not fit your requirements, please let us know via the `laminas-cache` [repository on GitHub](https://github.com/laminas/laminas-cache) and tell us more about your implementations. Maybe your addition should be part of our official adapter or could be provided as a dedicated Plugin instead.
```